### PR TITLE
More performance optimizations

### PR DESCRIPTION
**Pull request by dstepanov**

### CODE REVIEW

File: router/src/main/java/io/micronaut/web/router/UriRouteInfo.java [hunk 1]
```diff
@@ -51,9 +52,21 @@
...
     }
 
+    /**
+     * Match this route within the given URI and produce a {@link RouteMatch} if a match is found.
+     *
+     * @param uri The URI The URI
+     * @return A null or a {@link RouteMatch}
+     */
+    @Nullable
+    default UriRouteMatch<T, R> tryMatch(@NonNull  URI uri) {
```
Comment (yawkat) [inline router/src/main/java/io/micronaut/web/router/UriRouteInfo.java]:
```suggestion
    default UriRouteMatch<T, R> tryMatch(@NonNull URI uri) {
```

File: router/src/main/java/io/micronaut/web/router/UriRouteInfo.java [hunk 2]
```diff
@@ -51,9 +52,21 @@
      */
     @Override
     default Optional<UriRouteMatch<T, R>> match(URI uri) {
+
```
Comment (yawkat) [inline router/src/main/java/io/micronaut/web/router/UriRouteInfo.java]:
```suggestion
```

File: router/src/main/java/io/micronaut/web/router/UriRouteInfo.java [hunk 3]
```diff
@@ -63,6 +76,15 @@ default Optional<UriRouteMatch<T, R>> match(URI uri) {
...
     Optional<UriRouteMatch<T, R>> match(String uri);
 
+    /**
+     * Match this route within the given URI and produce a {@link RouteMatch} if a match is found.
+     *
+     * @param uri The URI
+     * @return A null or a {@link RouteMatch}
+     */
+    @Nullable
+    UriRouteMatch<T, R> tryMatch(@NonNull  String uri);
```
Comment (yawkat) [inline router/src/main/java/io/micronaut/web/router/UriRouteInfo.java]:
```suggestion
    UriRouteMatch<T, R> tryMatch(@NonNull String uri);
```

File: router/src/main/java/io/micronaut/web/router/DefaultRouter.java [hunk 4]
```diff
@@ -278,9 +290,20 @@ public <T, R> List<UriRouteMatch<T, R>> findAllClosest(@NonNull HttpRequest<?> r
...
+            UriRouteMatch match = route.tryMatch(path);
+            if (match != null) {
+                uriRoutes.add(match);
+            }
+        }
+        return uriRoutes;
+    }
+
+    private <T, R> List<UriRouteMatch<T, R>> toMatches(String path, UriRouteInfo<Object, Object>[] routes) {
+        List<UriRouteMatch<T, R>> uriRoutes = new ArrayList<>(routes.length);
```
Comment (yawkat) [inline router/src/main/java/io/micronaut/web/router/DefaultRouter.java:316]: can we optimize this using List.of, is routes.length commonly 1?

**Review by yawkat — APPROVED** [review]

File: router/src/main/java/io/micronaut/web/router/DefaultRouter.java (hunk 4 above)
Comment (dstepanov) [reply router/src/main/java/io/micronaut/web/router/DefaultRouter.java:316]: Added

**Review by graemerocher — APPROVED** [review]

**Pull request merged**